### PR TITLE
Individual CUDA object spilling  

### PR DESCRIPTION
**Pull request by madsbk**
This PR introduces a new _device host file_ that uses `ProxyObejct` to implement spilling of individual CUDA objects as opposed to the current host file, which spills entire keys. 

- [x] Implement spilling of individual objects
- [x] Handle task level aliasing
- [x] Handle shared device buffers
- [x] Write docs

To use, set `DASK_JIT_UNSPILL=True`

### Aliases at the task level 

Consider the following two tasks:
```python

def task1():  # Create list of dataframes
    df1 = cudf.DataFrame({"a": range(10)})
    df2 = cudf.DataFrame({"a": range(10)})
    return [df1, df2]

def task2(dfs):  # Get the second item
    return dfs[1]    
```
Running the two task on a worker we get something like:
```python

>>> data["k1"] = task1()
>>> data["k2"] = task2(data["k1"])
>>> data
{
    "k1": [df1, df2],
    "k2": df2,
}
```
Since the current implementation of spilling works on keys and handles each keys separately, it overestimate the device memory used: `sizeof(df)*3`. But even worse, if it decides to spill `k2` no device memory is freed since `k1` still holds a reference to `df2`!

The new spilling implementation fixes this issue by wrapping identical CUDA objects in a shared `ProxyObejct` thus in this case `df2` in both `k1` and `k2` will refer to the same `ProxyObejct`.

### Sharing device buffers

Consider the following code snippet:
```python

>>> data["df"] = cudf.DataFrame({"a": range(10)})
>>> data["grouped"] = shuffle_group(data["df"], "a", 0, 2, 2, False, 2)
>>> data["v1"] = data["grouped"][0]
>>> data["v2"] = data["grouped"][1]
```
In this case `v1` and `v2` are separate objects and are handled separately both in the current and the new spilling implementation. However, the `shuffle_group()` in cudf actually returns a single device memory buffer such that `v1` and `v2` points to the same underlying memory buffer. Thus the current implement will again overestimate the memory use and spill one of the dataframes without any effect.
The new implementation takes this into account when estimating memory usage and make sure that either both dataframes are spilled or none of them are.

cc. @beckernick, @VibhuJawa
xref: https://github.com/dask/distributed/issues/3756

### CODE REVIEW

**Comment by beckernick** [comment]
From our initial tests, this works exactly as we had all hoped. Memory is well behaved, and the performance impact is significant. Query results also pass the correctness checks.

Setup:
- 8 GPUs of a DGX-2 (GPUs 0-7)
- DEVICE_MEMORY_LIMIT="15GB"
- POOL_SIZE="30GB"
- TCP communication
- Reading parquet files comprising 2GB in-memory data chunks from the local /raid of the DGX-2

Q02 Standard: 300 seconds
Q02 Object Spilling: 85 seconds

Q03 Standard: 295 seconds
Q03 Object Spilling: 100 seconds

cc @quasiben @kkraus14

**Comment by quasiben** [comment]
Those are significant improvements!

**Comment by beckernick** [comment]
Some queries are failing during equality comparisons (ProxyObjects don't support equality ops). Looks like there's several other errors, but will need to distinguish between them and other issues.

For queries with which this succeeds, it's fantastic.

**Comment by beckernick** [comment]
There are no failures in the standard environment. Will be using this comment to track object spilling failures with @madsbk 

**Q01**
TypeError: '<' not supported between instances of 'ProxyObject' and 'ProxyObject'

Likely coming from the less than operation in the custom task https://github.com/rapidsai/gpu-bdb/blob/1dc201bdb4542213df265c87b21e8e989291cacc/tpcx_bb/queries/q01/tpcx_bb_query_01.py#L78

**Q05**
TypeError: Implicit conversion to a host NumPy array via __array__ is not allowed, To explicitly construct a GPU array, consider using cupy.asarray(...)
To explicitly construct a host array, consider using .to_array()

Likely somewhere in build_and_predict_model: https://github.com/rapidsai/gpu-bdb/blob/main/tpcx_bb/queries/q05/tpcx_bb_query_05.py#L75

**Q07**
TypeError: '>' not supported between instances of 'ProxyObject' and 'ProxyObject'

**Q10**
  File "cudf/_lib/merge.pyx", line 36, in cudf._lib.merge.merge_sorted
TypeError: Cannot convert ProxyObject to cudf._lib.table.Table

... others to be filled in as time permits ...

**Comment by jakirkham** [comment]
My guess is `ProxyObject`s will need to convert themselves back to real objects when performing those comparisons.

**Comment by jakirkham** [comment]
> My guess is `ProxyObject`s will need to convert themselves back to real objects when performing those comparisons.

Done in PR ( https://github.com/rapidsai/dask-cuda/pull/458 )

**Comment by madsbk** [comment]
@beckernick, I have updated your error log. Q01, q05, q07 should now work. Please re-add them if  they still fails for you

**Comment by beckernick** [comment]
👍  I see this also. Just for tracking sake, I'm going to re-do your comment edits as ~strikethrough~

**Comment by madsbk** [comment]
@beckernick, all queues seems to work now. Can you confirm that they also works for you?
Notice, you need `dask-master` from today, which include https://github.com/dask/dask/pull/6927

**Comment by madsbk** [comment]
### Test and evaluation of NVTabular's [Criteo/DLRM Preprocessing Benchmark](https://github.com/NVIDIA/NVTabular/blob/main/examples/dask-nvtabular-criteo-benchmark.py)

Running a 10-days data set on a on DXG-2 everything works and achieve a **1.48** times speedup. In order to force the use of _device-to-host_ memory spilling, I had to set the device limit to 6.5GB.

|   | Memory Peak | Runtime  |  Speedup  | 
|---|---|---|---|
| Old style spilling | 20 GB | 408 sec | 1 | 
| JIT unspill  | 17 GB  | 276 sec  | 1.48  | 

The commands, which depend on Dask `master`:
```
# Old style spilling
DASK_JIT_UNSPILL=False python NVTabular/examples/dask-nvtabular-criteo-benchmark.py --data-path /datasets/criteo/crit_orig_pq_10days --out-path output -d "8,9,10,11,12,13,14,15" --cats-on-device --cat-cache-high device --device-limit-frac 0.2

# JIT unspill
DASK_JIT_UNSPILL=True  python NVTabular/examples/dask-nvtabular-criteo-benchmark.py --data-path /datasets/criteo/crit_orig_pq_10days --out-path output -d "8,9,10,11,12,13,14,15" --cats-on-device --cat-cache-high device --device-limit-frac 0.2
```

Based on the results from the NVTabular and TPCx-BB workflows, I think this PR is ready for reviews.

File: dask_cuda/proxify_host_file.py
Comment (pentschev) [inline dask_cuda/proxify_host_file.py:63]:
```suggestion
    - When a proxy in the host file is deserialized
    - When a proxy in the host file is serialized
```

**Review by pentschev — CHANGES_REQUESTED** [review]
From a high-level, I think this PR looks good, I have a few proposed changes and questions below.

Overall the PR is very complex and difficult to keep track of all the important details when reviewing, so in all honesty I couldn't mentally validate if all the connections between different classes and various data types registrations are indeed correct, I think the best approach here is to indeed have this as an experimental feature (as currently proposed) for some type to allow people to validate the system functionality.

Thanks @madsbk for the huge effort you put into this problem!

File: dask_cuda/cuda_worker.py
Comment (madsbk) [reply dask_cuda/cuda_worker.py:195]: Yes, in this PR we will not support _any_ spilling to disk. I have removed `memory_limit` and `local_directory` to make this clear. I don't think `local_directory` is used for anything else than disk spilling?

I will be up to a future PR to implement disk spilling, which shouldn't be too difficult.

File: dask_cuda/tests/test_proxy.py
Comment (madsbk) [reply dask_cuda/tests/test_proxy.py:394]: Changed to `dask.array.tensordot()` and `dask.array.einsum()`

File: dask_cuda/tests/test_proxify_host_file.py
Comment (madsbk) [reply dask_cuda/tests/test_proxify_host_file.py:45]: I have change the test to use `dhf.proxies_tally` for the check.

File: dask_cuda/proxy_object.py
Comment (madsbk) [reply dask_cuda/proxy_object.py:236]: Changed to `_obj_pxy_is_serialized`

File: dask_cuda/proxy_object.py
Comment (madsbk) [reply dask_cuda/proxy_object.py:730]: Removed

**Comment by madsbk** [comment]
@pentschev thanks for the review, much appreciated. I have addressed all of your suggestions, I think :)

**Comment by pentschev** [comment]
Thanks @madsbk for addressing those. It seems that failing tests are legit, all three seem to fail with similar errors:

```python
dask_cuda/tests/test_proxy.py::test_proxy_object_parquet *** stack smashing detected ***: <unknown> terminated
Fatal Python error: Aborted
```

File: dask_cuda/cuda_worker.py
Comment (pentschev) [reply dask_cuda/cuda_worker.py:195]: My understanding was that `local_directory` was used for more than just spilling, so I went on a hunt. To be fair, it's still unclear to me even where in the code `local_directory` is really used, I found one other case where it's used though, when you upload files, see for example https://github.com/dask/distributed/blob/607cfd2ce00edd44c99da3273de0763a426dda7d/distributed/tests/test_worker.py#L176-L202 . That isn't to say this is the only other case, but I couldn't confirm nor deny whether there are other cases besides spilling and file uploading. Maybe @quasiben or @jakirkham would know.

**Review by pentschev — APPROVED** [review]
Thanks for checking the failing tests @madsbk . To the extent I was capable of verifying this PR, everything looks good to me now. I'm approving but will wait until tomorrow for merging to give others a chance to review it as well.

**Pull request merged**
